puresasl/mechanisms.py

```diff
@@ -187,6 +190,56 @@ def process(self, challenge=None):
     def dispose(self):
         self.password = None
 
+    def wrap(self, outgoing):
+        if self.qop == 'auth-int' or self.qop == 'auth-conf':
+            raise Exception('{0} QoP not supported for CRAM-MD5'.format(self.qop))
```

Comment:
If it's primarily to add a notion of supported QOPs to mechanisms (which seems to be the case here), we could just add a hard-coded list to each mechanism and check that up front.